man/xml2tlk.1
.Dd June 17, 2020
.Dt XML2TLK 1
.Os
.Sh NAME
.Nm xml2tlk
.Nd XML to BioWare TLK converter
.Sh SYNOPSIS
.Nm xml2tlk
.Op Ar options
.Op Ar input_file
.Ar output_file
.Sh DESCRIPTION
.Nm
converts XML files created by the
.Xr tlk2xml 1
tool back into the BioWare TLK format.
For a more in-depth description of TLK files,
please see the man page for the
.Xr tlk2xml 1
tool.
Also note that currently, only the non-GFF versions, V3.0 and V4.0,
can be created by xml2tlk.
.Pp
The format of the input XML is pretty simple and straight-forward.
.Bd -literal
<?xml version="1.0" encoding="utf-8" standalone="yes"?>
<tlk language="0">
  <string id="1">Continue</string>
  <string id="2" sound="hello">Well hello there!</string>
  <string id="3" sound="bye" soundlength="0.5">Bye!</string>
  <string id="4" soundid="23">Who are you?</string>
</tlk>
.Ed
.Pp
The root element is
.Dq tlk ,
and it can have an optional language property.
That language ID can also be given on the command line,
and it then overrides the one in the input XML.
When creating a,
versions V3.0 and V4.0 need a language ID, while versions V0.2
and V0.5 ignore the language ID.
.Pp
Each child tag of the root element has to be a
.Dq string
element, and each
.Dq string
element requires an
.Dq id
property.
The ID is the string reference (StrRef) for the text line,
and the contents of the
.Dq string
element is the text itself.
The whole file has to be UTF-8 encoded.
.Pp
Version V3.0 allows the following extra properties on a
.Dq string :
.Dq sound
(a resource reference of a voice-over for this line, \(<=\ 16
characters),
.Dq soundlength
(a floating point number denoting the length of the sound file in seconds),
.Dq volumevariance
(unused by the games) and
.Dq pitchvariance
(unused by the games).
.Pp
Version V4.0 allows the extra property
.Dq soundid
on a
.Dq string ,
which is a numerical references to a voice-over line.
.Pp
Versions V0.2 and V0.5 do not allow any extra properties.
.Pp
Because TLK files contain localized string data, it is important
to know the encoding of those strings.
Unfortunately, the TLK files do not contain information about the encoding.
Version 3.0 and 4.0 contain a language identifier,
but the meaning of that varies between games.
V0.2 and V0.5 even lack those completely.
However, due to the Huffman-nature of V0.5 strings, the encoding
there is fixed to little-endian UTF-16, and strings in V0.2 files
are also usually in little-endian UTF-16 (with the exceptions of
files found in the Nintendo DS game
.Em Sonic Chronicles: The Dark Brotherhood ) .
To manually select the encoding, this tool provides a wide range
command line options for various encodings.
.Pp
To start, the game this TLK is for needs to be specified. If the
XML file contains a language ID (which is the case for XML created
from TLK files of every game except
.Em Sonic Chronicles: The Dark Brotherhood ) ,
.Nm
then writes strings in the encoding fitting for that game with this
language. Alternatively, the language can be overwritten as well
as the encoding itself.
.Pp
Likewise, the version of TLK file to create is by default taken from
the game parameter, i.e. a TLK file that can be used with this particular
game is written. This can also be overwritten with the appropriate flag.
Please note that as of yet, the versions used by the games
.Em Sonic Chronicles: The Dark Brotherhood ,
.Em Dragon Age: Origins
and
.Em Dragon Age II
are not supported. There, a version flag needs to be given (but the
created file will not be readable by the game itself).
.Sh OPTIONS
.Bl -tag -width xxxx -compact
.It Fl h
.It Fl Fl help
Show a help text and exit.
.It Fl Fl version
Show version information and exit.
.It Fl 3
.It Fl Fl version30
Write a V3.0 TLK file.
.It Fl 4
.It Fl Fl version40
Write a V4.0 TLK file.
.It Fl l Ar id
.It Fl Fl language Ar id
Override the TLK language ID.
.It Fl Fl cp1250
Write strings as Windows CP-1250.
.Pp
Eastern European, Latin alphabet.
.It Fl Fl cp1251
Write strings as Windows CP-1251.
.Pp
Eastern European, Cyrillic alphabet.
.It Fl Fl cp1252
Write strings as Windows CP-1252.
.Pp
Western European, Latin alphabet.
.It Fl Fl cp932
Write strings as Windows CP-932.
.Pp
Japanese, extended Shift-JIS.
.It Fl Fl cp936
Write strings as Windows CP-936.
.Pp
Simplified Chinese, extended GB2312 with GBK codepoints.
.It Fl Fl cp949
Write strings as Windows CP-949.
.Pp
Korean, similar to EUC-KR.
.It Fl Fl cp950
Write strings as Windows CP-950.
.Pp
Traditional Chinese, similar to Big5.
.It Fl Fl utf8
Write strings as UTF-8.
.It Fl Fl utf16le
Write strings as little-endian UTF-16.
.It Fl Fl utf16be
Write strings as big-endian UTF-16.
.It Fl Fl nwn
Write strings in an encoding appropriate for
.Em Neverwinter Nights .
.It Fl Fl nwn2
Write strings in an encoding appropriate for
.Em Neverwinter Nights 2 .
.It Fl Fl kotor
Write strings in an encoding appropriate for
.Em Knights of the Old Republic .
.It Fl Fl kotor2
Write strings in an encoding appropriate for
.Em Knights of the Old Republic II .
.It Fl Fl jade
Write strings in an encoding appropriate for
.Em Jade Empire .
.It Fl Fl witcher
Write strings in an encoding appropriate for
.Em The Witcher .
.It Fl Fl dragonage
Write strings in an encoding appropriate for
.Em Dragon Age: Origins .
.It Fl Fl dragonage2
Write strings in an encoding appropriate for
.Em Dragon Age II .
.El
.Bl -tag -width xxxx -compact
.It Ar input_file
The XML file to convert.
If no input file is specified, the XML data is read from
.Dv stdin .
The encoding of the XML stream must always be UTF-8.
.It Ar output_file
The TLK file will be written there.
.El
.Sh EXAMPLES
Convert
.Pa file1.xml
into a V3.0 CP-1252 TLK file:
.Pp
.Dl $ xml2tlk --version30 --cp1252 file1.xml file2.tlk
.Pp
Convert
.Pa file1.xml
into a V4.0 UTF-8 TLK file and override the language ID:
.Pp
.Dl $ xml2tlk --version40 --utf8 --language 1 file1.xml file2.tlk
.Pp
Convert
.Pa file1.xml
into a V3.0 TLK file from Neverwinter Nights:
.Pp
.Dl $ xml2tlk --version30 --nwn file1.xml file2.tlk
.Pp
Convert the UTF-8 TLK
.Pa file1.tlk
into an XML file on
.Dv stdout
with
.Xr tlk2xml 1 ,
modify it using
.Xr sed 1
and write the result back into a TLK:
.Bd -literal -offset xxxxxx
$ tlk2xml --utf8 file1.tlk | sed -e 's/gold/candy/g' | xml2tlk \e
  --utf8 --version30 file2.tlk
.Ed
.Sh SEE ALSO
.Xr tlk2xml 1 ,
.Xr xml2ssf 1
.Pp
More information about the xoreos project can be found on
.Lk https://xoreos.org/ "its website"
.Ns .
.Sh AUTHORS
This program is part of the xoreos-tools package, which in turn is
part of the xoreos project, and was written by the xoreos team.
Please see the
.Pa AUTHORS
file for details.


NEWS
Changes in xoreos-tools version 0.0.2
=====================================

This is the first official release of xoreos-tools, together with xoreos.

Included are:

- Tools for handling resource archives
  - unkeybif
  - unerf
  - unrim
  - unherf
  - unnds
  - desmall
- Tools converting basic resource formats into human-readable form
  - gff2xml
  - tlk2xl
  - convert2da
- Tools converting graphics formats into TGA
  - xoreostex2tga
  - unnsbtx
  - nbfs2tga
  - ncgr2tga
  - cbgt2tga
  - cdpth2tga


src/xml2tlk.cpp
/* xoreos-tools - Tools to help with xoreos development
 *
 * xoreos-tools is the legal property of its developers, whose names
 * can be found in the AUTHORS file distributed with this source
 * distribution.
 *
 * xoreos-tools is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 3
 * of the License, or (at your option) any later version.
 *
 * xoreos-tools is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with xoreos-tools. If not, see <http://www.gnu.org/licenses/>.
 */

/** @file
 *  Tool to convert XML files back into TLK.
 */

#include <cstring>
#include <cstdio>

#include "src/version/version.h"

#include "src/common/scopedptr.h"
#include "src/common/ustring.h"
#include "src/common/util.h"
#include "src/common/strutil.h"
#include "src/common/error.h"
#include "src/common/platform.h"
#include "src/common/readfile.h"
#include "src/common/writefile.h"
#include "src/common/stdinstream.h"
#include "src/common/encoding.h"
#include "src/common/cli.h"

#include "src/aurora/types.h"
#include "src/aurora/language.h"

#include "src/xml/tlkcreator.h"

#include "src/util.h"

bool parseCommandLine(const std::vector<Common::UString> &argv, int &returnValue,
                      Common::UString &inFile, Common::UString &outFile,
                      Common::Encoding &encoding, Aurora::GameID &game,
                      XML::TLKCreator::Version &version, uint32 &language);

void createTLK(const Common::UString &inFile, const Common::UString &outFile, Common::Encoding encoding,
               XML::TLKCreator::Version &version, uint32 &language);

int main(int argc, char **argv) {
	initPlatform();

	try {
		std::vector<Common::UString> args;
		Common::Platform::getParameters(argc, argv, args);

		Common::Encoding encoding = Common::kEncodingInvalid;
		Aurora::GameID   game     = Aurora::kGameIDUnknown;

		XML::TLKCreator::Version version = XML::TLKCreator::kVersionInvalid;
		uint32 language = Aurora::kLanguageInvalid;

		int returnValue = 1;
		Common::UString inFile, outFile;

		if (!parseCommandLine(args, returnValue, inFile, outFile, encoding, game, version, language))
			return returnValue;

		LangMan.declareLanguages(game);

		createTLK(inFile, outFile, encoding, version, language);
	} catch (...) {
		Common::exceptionDispatcherError();
	}

	return 0;
}

bool parseCommandLine(const std::vector<Common::UString> &argv, int &returnValue,
                      Common::UString &inFile, Common::UString &outFile,
                      Common::Encoding &encoding, Aurora::GameID &game,
                      XML::TLKCreator::Version &version, uint32 &language) {

	using Common::CLI::NoOption;
	using Common::CLI::kContinueParsing;
	using Common::CLI::Parser;
	using Common::CLI::ValGetter;
	using Common::CLI::ValAssigner;
	using Common::CLI::makeEndArgs;
	using Common::CLI::makeAssigners;
	using Common::Encoding;
	using Aurora::GameID;
	std::vector<Common::UString> args;
	NoOption inFileOpt(false, new ValGetter<Common::UString &>(inFile, "input files"));
	NoOption outFileOpt(true, new ValGetter<Common::UString &>(outFile, "output files"));
	Parser parser(argv[0], "XML to BioWare TLK converter",
	              "If no input file is given, the input is read from stdin.\n\n"
	              "One of the flags to identify the game this TLK is to be used is mandatory,\n"
	              "as is a language ID unless the XML file specifies one (in which case, the\n"
	              "--language parameter can be used to override it).\n\n"
	              "By default, the version of TLK fitting for the specified game and the\n"
	              "encoding fitting for the game and language is used. These choice can be\n"
	              "overwritten by using the appropriate encoding and --version flags.\n",
	              returnValue, makeEndArgs(&inFileOpt, &outFileOpt));

	encoding = Common::kEncodingInvalid;
	game     = Aurora::kGameIDUnknown;
	version  = XML::TLKCreator::kVersionInvalid;
	language = Aurora::kLanguageInvalid;

	parser.addSpace();
	parser.addOption("version30", '3', "Write a V3.0 TLK file", kContinueParsing,
	                 makeAssigners(new ValAssigner<XML::TLKCreator::Version>
	                 (XML::TLKCreator::kVersion30, version)));
	parser.addOption("version40", '4', "Write a V4.0 TLK file", kContinueParsing,
	                 makeAssigners(new ValAssigner<XML::TLKCreator::Version>
	                 (XML::TLKCreator::kVersion40, version)));

	parser.addOption("language", 'l', "Override the TLK language ID", kContinueParsing,
	                 new ValGetter<uint32_t &>(language, "id"));

	parser.addSpace();
	parser.addOption("cp1250", "Write TLK strings as Windows CP-1250", kContinueParsing,
	                 makeAssigners(new ValAssigner<Encoding>(Common::kEncodingCP1250, encoding),
	                 new ValAssigner<GameID>(Aurora::kGameIDUnknown, game)));
	parser.addOption("cp1251", "Write TLK strings as Windows CP-1251", kContinueParsing,
	                 makeAssigners(new ValAssigner<Encoding>(Common::kEncodingCP1251, encoding),
	                 new ValAssigner<GameID>(Aurora::kGameIDUnknown, game)));
	parser.addOption("cp1252", "Write TLK strings as Windows CP-1252", kContinueParsing,
	                 makeAssigners(new ValAssigner<Encoding>(Common::kEncodingCP1252, encoding),
	                 new ValAssigner<GameID>(Aurora::kGameIDUnknown, game)));
	parser.addOption("cp932", "Write TLK strings as Windows CP-932", kContinueParsing,
	                 makeAssigners(new ValAssigner<Encoding>(Common::kEncodingCP932, encoding),
	                 new ValAssigner<GameID>(Aurora::kGameIDUnknown, game)));
	parser.addOption("cp936", "Write TLK strings as Windows CP-936", kContinueParsing,
	                 makeAssigners(new ValAssigner<Encoding>(Common::kEncodingCP936, encoding),
	                 new ValAssigner<GameID>(Aurora::kGameIDUnknown, game)));
	parser.addOption("cp949", "Write TLK strings as Windows CP-949", kContinueParsing,
	                 makeAssigners(new ValAssigner<Encoding>(Common::kEncodingCP949, encoding),
	                 new ValAssigner<GameID>(Aurora::kGameIDUnknown, game)));
	parser.addOption("cp950", "Write TLK strings as Windows CP-950", kContinueParsing,
	                 makeAssigners(new ValAssigner<Encoding>(Common::kEncodingCP950, encoding),
	                 new ValAssigner<GameID>(Aurora::kGameIDUnknown, game)));
	parser.addOption("utf8", "Write TLK strings as UTF-8", kContinueParsing,
	                 makeAssigners(new ValAssigner<Encoding>(Common::kEncodingUTF8, encoding),
	                 new ValAssigner<GameID>(Aurora::kGameIDUnknown, game)));
	parser.addOption("utf16le", "Write TLK strings as little-endian UTF-16", kContinueParsing,
	                 makeAssigners(new ValAssigner<Encoding>(Common::kEncodingUTF16LE, encoding),
	                 new ValAssigner<GameID>(Aurora::kGameIDUnknown, game)));
	parser.addOption("utf16be", "Write TLK strings as big-endian UTF-16", kContinueParsing,
	                 makeAssigners(new ValAssigner<Encoding>(Common::kEncodingUTF16BE, encoding),
	                 new ValAssigner<GameID>(Aurora::kGameIDUnknown, game)));
	parser.addSpace();
	parser.addOption("nwn", "Use Neverwinter Nights encodings", kContinueParsing,
	                 makeAssigners(new ValAssigner<Encoding>(Common::kEncodingInvalid, encoding),
	                 new ValAssigner<GameID>(Aurora::kGameIDNWN, game)));
	parser.addOption("nwn2", "Use Neverwinter Nights 2 encodings", kContinueParsing,
	                 makeAssigners(new ValAssigner<Encoding>(Common::kEncodingInvalid, encoding),
	                 new ValAssigner<GameID>(Aurora::kGameIDNWN2, game)));
	parser.addOption("kotor", "Use Knights of the Old Republic encodings", kContinueParsing,
	                 makeAssigners(new ValAssigner<Encoding>(Common::kEncodingInvalid, encoding),
	                 new ValAssigner<GameID>(Aurora::kGameIDKotOR, game)));
	parser.addOption("kotor2", "Use Knights of the Old Republic II encodings", kContinueParsing,
	                 makeAssigners(new ValAssigner<Encoding>(Common::kEncodingInvalid, encoding),
	                 new ValAssigner<GameID>(Aurora::kGameIDKotOR2, game)));
	parser.addOption("jade", "Use Jade Empire encodings", kContinueParsing,
	                 makeAssigners(new ValAssigner<Encoding>(Common::kEncodingInvalid, encoding),
	                 new ValAssigner<GameID>(Aurora::kGameIDJade, game)));
	parser.addOption("witcher", "Use The Witcher encodings", kContinueParsing,
	                 makeAssigners(new ValAssigner<Encoding>(Common::kEncodingInvalid, encoding),
	                 new ValAssigner<GameID>(Aurora::kGameIDWitcher, game)));
	parser.addOption("dragonage", "Use Dragon Age encodings", kContinueParsing,
	                 makeAssigners(new ValAssigner<Encoding>(Common::kEncodingInvalid, encoding),
	                 new ValAssigner<GameID>(Aurora::kGameIDDragonAge, game)));
	parser.addOption("dragonage2", "Use Dragon Age II encodings", kContinueParsing,
	                 makeAssigners(new ValAssigner<Encoding>(Common::kEncodingInvalid, encoding),
	                 new ValAssigner<GameID>(Aurora::kGameIDDragonAge2, game)));

	if (!parser.process(argv))
		return false;

	if (game == Aurora::kGameIDUnknown) {
		parser.usage();
		returnValue = Common::CLI::kEndFail;
		return false;
	}

	if (version == XML::TLKCreator::kVersionInvalid) {
		switch (game) {
			case Aurora::kGameIDNWN:     XOREOS_FALLTHROUGH;
			case Aurora::kGameIDNWN2:    XOREOS_FALLTHROUGH;
			case Aurora::kGameIDKotOR:   XOREOS_FALLTHROUGH;
			case Aurora::kGameIDKotOR2:  XOREOS_FALLTHROUGH;
			case Aurora::kGameIDWitcher:
				version = XML::TLKCreator::kVersion30;
				break;
			case Aurora::kGameIDJade:
				version = XML::TLKCreator::kVersion40;
				break;
			case Aurora::kGameIDSonic:
			case Aurora::kGameIDDragonAge:
			case Aurora::kGameIDDragonAge2:
				std::printf("Specified game uses unsupported TLK version.\n");
				returnValue = Common::CLI::kEndFail;
				return false;

			default:
				break;
		}
	}

	return true;
}

void createTLK(const Common::UString &inFile, const Common::UString &outFile, Common::Encoding encoding,
               XML::TLKCreator::Version &version, uint32 &language) {

	Common::WriteFile tlk(outFile);
	Common::ScopedPtr<Common::ReadStream> xml(openFileOrStdIn(inFile));

	XML::TLKCreator::create(tlk, *xml, version, encoding, inFile, language);

	tlk.flush();
	tlk.close();
}
